File: week-12/rimrim990/실패율.py
def solution(N, stages):
  clear = [0] * (N+3)
  reached = [0] * (N+3)

  for stage in stages:
    # 스테이지를 클리어한 사용자의 수
    clear[1] += 1
    clear[stage] -= 1

    # 스테이지에 도달한 사용자의 수
    reached[1] += 1
    reached[stage+1] -= 1

  # 누적합
  fails = []
  for i in range(1, N+1):
    clear[i] += clear[i-1]
    reached[i] += reached[i-1]

    if reached[i]:
      val = (reached[i]-clear[i]) / reached[i]
    else:
      val = 0

    fails.append((val, i))

  fails.sort(reverse=True, key=lambda x: (x[0], -x[1]))
  return list(map(lambda x: x[1], fails))

File: week-12/dahyen0o/실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        final double[][] results = new double[N][2];
        
        for(int stage = 1; stage <= N; stage++) {
            int fail = 0;
            int total = stages.length;
            
            for(int st : stages) {
                if(st == stage) fail++;
                else if(st < stage) total--;
            }
            
            results[stage - 1][0] = stage;
            if(total > 0) results[stage - 1][1] = fail / (double) total;
        }
        
        Arrays.sort(results, new Comparator<double[]>() {
            @Override
            public int compare(double[] arr1, double[] arr2) {
                if(Double.compare(arr2[1], arr1[1]) == 0) {
                    return Double.compare(arr1[0], arr1[0]);
                }
                return Double.compare(arr2[1], arr1[1]);
            }
        });
        
        final int[] answer = new int[N];
        for(int n = 0; n < N; n++) {
            answer[n] = (int) results[n][0];
        }
        return answer;
    }
}
